### ISSUE TITLE
CorpseRebord 2.13 - spawns cows upon death

### ISSUE THREAD

**Issue by smmmadden** [egold555/CorpseReborn#57]
I'm an admin, OP and I spawned some creepers to kill me and then see how this works.  I died, but when I did I heard cows mooing.  I rejoined the server, went to my death spot and found about 30 cows where there once was a hole from the creepers and my corpse (though I couldn't see it clearly due to all the cows).
![image](https://user-images.githubusercontent.com/10309346/44620421-d7c36100-a861-11e8-94af-c8fa8dce7668.png)
![image](https://user-images.githubusercontent.com/10309346/44620425-f32e6c00-a861-11e8-9c9c-c632fe30c340.png)
This is using a Spigot 1.13 build from today (20 minutes to be exact).  When I tried to kill the cows, thats when the errors in the console/logs showed up.

**Comment by smmmadden** [comment]
CreeperHeal repaired the hole in the grown, but all the cows are still there when I exited and rejoined. :-(
![image](https://user-images.githubusercontent.com/10309346/44620449-751e9500-a862-11e8-93d3-5da8b4f60979.png)

**Comment by smmmadden** [comment]
I killed them all as they were still spawning after shutting down the server and removing the CorpseReborn plugin.  All said and done:
![image](https://user-images.githubusercontent.com/10309346/44620544-3a1d6100-a864-11e8-89b3-ffc294122172.png)

What's strange is the name given to all the cows: CRHitbox so I looked at the config.yml and saw this entry:
#Due to a Minecraft quirk with how the corpses work, the hitboxes were not normal player hitboxes.
#I have now made it so a invisible Cows spawn that covers the whole corpse.
#When you click on the invisible Cow it does the same thing as clicking on the corpse.
#Set to false to use the old glitchy hitboxes.
new-hitboxes: true
No idea why cows need to be spawned to begin with invisible or otherwise.

**Comment by egold555** [comment]
Hi. Thanks for letting me know about this. Only 1 invidable cow is suppost to be spawned at the corpse due to how players who are in bed do not have a player hitbox. This is why I use a cow to detect when you click on the corpse. I will download the latest version and see what is broken. Could you also do "/cr dump" and send me the pastebin link? If that doesnt work /version will do

**Comment by egold555** [comment]
I can not replicate with just CorpseReborn by it's self. I am wondering if some other plugin that you are using is causing some strange event issues. I will be happy to test using the plugins you have from the cr paste

**Comment by egold555** [comment]
Done a lot more testing and found that there is a edge case where killing and clicking on cows would cause a null pointer, but this still doesn't explain the massive amount of cows.

**Comment by smmmadden** [comment]
>plugins
[14:30:12 INFO]: Plugins (43): ColorCodes, dynmap, LuckPerms, WorldEdit, OpenInv, PlaceholderAPI, Realistic-Swimming, PacketListenerApi, WorldBorder, Votifier, AutoBroadcasterPlus, ProtocolLib, Socket4MC, BungeeServerSigns, VotingPlugin, EditableSign, HorseTpWithMe, Ridables, CommandSync, AutoTorch, Links, Vault, BungeeTabListPlus, MassiveCore, Essentials, CreativeGates, HeadDatabase, LWC, QuickShop, SaneEconomy, SecurityVillagers, MythicMobs, Residence, EssentialsProtect, EssentialsChat, EssentialsGeoIP, EssentialsAntiBuild, CreeperHeal, MassiveBooks, EssentialsExtras, EssentialsSpawn, Jobs, SelectionVisualizer

I was able to do it again and this time instead of leaving and respawning I sat on the screen and the screen where it said I died was glitching (shaking repeatedly and cows started spawning as seen in below timelapse.  cr dump did nothing though.
![image](https://user-images.githubusercontent.com/10309346/44621426-53c6a480-a874-11e8-91bd-44473fa2c9c4.png)
![image](https://user-images.githubusercontent.com/10309346/44621430-60e39380-a874-11e8-85bf-c45912289c0a.png)
![image](https://user-images.githubusercontent.com/10309346/44621432-6640de00-a874-11e8-9a84-c7d3c8c11838.png)
![image](https://user-images.githubusercontent.com/10309346/44621434-6ccf5580-a874-11e8-913d-3c6a781aa5b5.png)
![image](https://user-images.githubusercontent.com/10309346/44621436-71940980-a874-11e8-955a-961f7dbb6086.png)

**Comment by smmmadden** [comment]
and they keep on spawning from the same spot - I could see cow after cow come out from the corner block and killall wouldn't kill them.  Shutting down server, removing CR and starting it up I'm also not able to kill the cows with killall or any command other than without the plugin and then killing them one by one. :-(

**Comment by egold555** [comment]
Hm very strange. Still looking into this. Ill hoepfully get a bigger chunk of time to look into why this is happning tomorrow.

**Comment by smmmadden** [comment]
Thanks, I appreciate it.  This is one of those resources that I'd like to offer on my 1.13 servers once everything settles down and they are production ready. :-)

**Comment by egold555** [comment]
Just updated the plugin to hopefully fix this bug. I still have not been able to replicate it in my testing environment. Let me know if the update fixed this issue or not.

**Comment by smmmadden** [comment]
sure thing, will test again and see if that does it.  Thanks!

**Comment by smmmadden** [comment]
Just tested with 1.13 (still waiting on a couple other fixes for 1.13.1 to test there, but it works!)
I blew myself up with some creepers
![image](https://user-images.githubusercontent.com/10309346/44751977-abb91180-aae7-11e8-8854-bc4034a5544a.png)
Respawned and returned to death point
![image](https://user-images.githubusercontent.com/10309346/44752015-c1c6d200-aae7-11e8-9d7b-ea0b970aff85.png)
And was able to right click to loot the corpse.  Once the loot was collected, the corpse disappears.
![image](https://user-images.githubusercontent.com/10309346/44752030-cdb29400-aae7-11e8-8d00-ee74dc39ff2d.png)

Thanks for the quick turnaround and fixing it.  :-)  Will be posting positive review shortly.

**Comment by smmmadden** [comment]
Just tested with 1.13.1 and the issue is back again with the spawning cows. :-(

**Comment by egold555** [comment]
This makes no sense on why this happens and why I can not replicate this issue. For the time being I guess you can turn off the cow hitbox and use the alternitive, the cow just makes it way more easy to click on the corpse. Ill keep trying to replicate this because this is a big bug

**Comment by egold555** [comment]
Would there be a way for you to send me a zip of all your plugins and plugin configurations so I can boot up a test server and debug more? I have downloaded all the plugins from the "/cr dump" and for the life of me can not get this issue to happen for me.

If you don't feel comfortable with sending me your plugins folder and configs I totally understand, I can keep trying to replicate it with the plugins I have downloaded from the dump file.

If you wish to send them, please just email me @ eric@golde.org, or if its a google drive link, ericgolde555@gmail.com .

**Comment by egold555** [comment]
also, if your using a custom build of spigot, that would be handy for testing

**Comment by smmmadden** [comment]
Sure, I'll package it up and put it up on my site so you can download it.  I really appreciate it, I know it can be frustrating when it isn't reproducible easily.  That's software for ya and after 35 years in this field, nothing surprises me.

**Comment by smmmadden** [comment]
Just sent email to you at eric@golde.org so you can download directly from my site.  Thanks again!
-Steve

**Comment by egold555** [comment]
Alright, good ish news, I am able to reproduce this bug with the files you have sent me. Now comes the potentially challenging part of finding why this is occurring. Also, apparently /kill @e command does not work with the hitboxes (I should add a command to remove bugged cows or something)

Thanks again for sending your files!

**Comment by egold555** [comment]
I found the bug! The plugin Rideables is doing some strange stuff with cows. If you disable the cow entity in the ridables config everything works as expected

**Comment by smmmadden** [comment]
That's great news!  What's odd is using Spigot it doesn't behave.  But I just downloaded and used PaperSpigot-187 (1.13.1) and I had no issues at all.  :-)    So it looks like a combination of Ridables and Spigot?

**Comment by egold555** [comment]
Hm Very very strange, thanks for pointing this out. Ill take a look.

**Comment by egold555** [comment]
From my bit of debugging, Ridables sets the isValid boolean on entity for all the cows to false when it is enabled. I think the easiest fix is to just not use that check and write my own check that can not be modified by any other plugins.

**Comment by egold555** [comment]
I can not find any references in the plugin to see if its valid, I am wondering if spigot is doing something

**Comment by BillyGalbreath** [comment]
Let me know if there's something I need to change in Ridables.I don't see this isValid method/field you speak of, though.

https://github.com/BillyGalbreath/Ridables/blob/master/src/main/java/net/pl3x/bukkit/ridables/entity/EntityRidableCow.java

**Comment by egold555** [comment]
Hi Billy,

Turns out that is not the problem. Looking in to your plugin some more, you kill the old entity cow and spawn a new one. https://github.com/BillyGalbreath/Ridables/blob/master/src/main/java/net/pl3x/bukkit/ridables/listener/RideListener.java#L151

I can not figure out a fix on my end but a quick one I can see in your plugin is if the cow's name is "CRHitbox" to just return and not spawn the rideable cow. I know this is kind of a hacky way and I am open if you have any other suggestions on ways I can fix it on my end.

**Comment by BillyGalbreath** [comment]
Ridables will only kill and replace non ridable entities. This means you are creating a new cow using nms (new EntityCow()) somewhere. Instead, if you use the Bukkit API to soawna cow it will be of ridable type (I injected the ridable type into the registry onLoad). Or, if you require nms methods right away (before world.addEntity()), get the function from the EntityTypes to create a new entity.

I wish I could give better examples, but I'm currently at work and won't be at a PC until earliest of Thursday (~36 hours from now).

**Comment by egold555** [comment]
I am confused about the whole PaperSpigot thing as well. 

@BillyGalbreath In order to prevent the cow from being rideable, would I not spawn the cow using the bukkit method and spawn it using NMS? Little bit confused what you stated above

**Comment by BillyGalbreath** [comment]
You want it to be ridable (EntityRidableCow), otherwise my plugin will replace it. You don't have to worry about the specifics of that, though. You can either spawn a cow using the Bukkit API or use the Function from nms EntityTypes.COW (use .apply() on the Function is similar to new EntityCow() but you aren't bothered by the specifics of which class/type it is).

EntityCow cow = EntityTypes.COW.aT.apply();

I can't remember if the Function is public or not, so may need reflectiin to access it. This will give you a new EntityCow instance (or EntityRidableCow in this case) that's not added to the world yet.

Or, if you don't need to make changes before it's added to the world you can do

EntityCow cow = EntityTypes.a(world);

This will do the same as above, but also add it to the world for you.

**Comment by egold555** [comment]
Alright. Ill see what I can implement. It will be a bit of work because currently the cow spawning code is not per version, its shared between all the NMS libraries. I also need to make sure that if your plugin is not on the server, then I spawn a normal cow.

**Comment by BillyGalbreath** [comment]
Right, using EntityTypes or the Bukkit API will give you whatever cow is in the registry. No need to check for specific plugins or whatnot ;)

**Comment by egold555** [comment]
Really? I am new to this type of registry overriding. I'll take a look at it. If this is true, I guess I am still confused on how this should all work out. So I spawn a cow, and your plugin removes it/ So what would I need to check for or implement to make sure your plugin doesn't remove it?

**Comment by BillyGalbreath** [comment]
According to that (https://github.com/egold555/CorpseReborn/blob/master/src/org/golde/bukkit/corpsereborn/nms/NmsBase.java#L98) you are using the API to spawn it in.

Hmm hmm hmm

Let me think about this for a minute.. this explains the difference between Spigot and Paper.

**Comment by egold555** [comment]
Currently I do not officially support PaperSpigot. It has always caused my plugin problems in the past lol.

I am happy to make a pull request for this but I think if on line 150 (https://github.com/BillyGalbreath/Ridables/blob/master/src/main/java/net/pl3x/bukkit/ridables/listener/RideListener.java#L150) if there was a check to see if the custom name if the entity is CRHitbox, and if so just return out of the event handler that handels weather or not to kill the mob or not. I haven't cloned your repo and tested it yet, but from looking at it I think it will work?

**Comment by egold555** [comment]
Its late at night so I would not be surprised if I am completely incorrect about that though

**Comment by BillyGalbreath** [comment]
I think all I have to do is ensure I copy over all the potion effect and nbt data from the old entity to the new one. If that doesn't work I'll had a special edge case to check for the custom name ^_^

**Comment by egold555** [comment]
Sounds good. Let me know if I can do anything to help :)

**Comment by smmmadden** [comment]
thank you both for figuring out what and where the conflict was - this was very enlightening and darn confusing at the same time. :-)  If we find that removing all these CRHitbox named cows isn't possible without a lot of work, I can reset the world as it is just a test world for running through all the testing for the plugins in 1.13 and 1.13.1.  Obviously not ideal, but I can offer that up once we think we've got it resolved in the next updates.  Thanks again Eric and Billy for your patience and time in all this!  -Steve

**Comment by BillyGalbreath** [comment]
@smmmadden you should be able to /kill the cows after a restart ;)

**Comment by smmmadden** [comment]
yeah, tried that too but when I use /minecraft:kill @e[type=cow] or [type=minecraft:cow] it highlights/hits the cows in red, but they don't despawn.  Repeated use of the same command does the same and leaves the CRHitbox cows in place.

**Comment by egold555** [comment]
I believe that is a issue on my side. You use to be able to kill them but with 1.13 I am wonder if the damage event now gets called with /kill ? More testing is required but I feel like my hitbox cows should obey the /kill command and just respawn the ones under the corpses.

**Comment by egold555** [comment]
#58 Just created this because its a bug, but unrelated to this issue.

**Comment by smmmadden** [comment]
Thanks!

**Comment by BillyGalbreath** [comment]
Issue is resolved https://github.com/BillyGalbreath/Ridables/commit/e8b4bdd2a7efc598d308a2aba9cb63c1faee80e6

\o/ ^_^

**Comment by smmmadden** [comment]
Hi Eric,

 

Just did some testing for Billy and Ridables for both Spigot and PaperSpigot 1.13.1.

 

With his cows entity enabled and disabled, when I died – no cows are spawned in both servers (Spigot & Paper)

/minecraft:kill @e[type=cow] or trying to kill with sword have no effect on CRHitbox cows.  In fact what did work previously was spawning some creepers while in god mode and that killed the cows.  Unfortunately, that option doesn’t work so I have no way to remove the previously spawned cows.  Any help on that would be great!

 

Also, a minor glitch is that when the corpse is left laying down, we should check for a solid block right?  Here is what happened when I did near water.



 

Technically, water is a block but wondering if there is anything to do to make sure it is on solid ground or float on/in water if that’s appropriate?  Again, minor nitpick. :)

 

Steve
